pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java

```diff
@@ -141,139 +233,116 @@ public void onInstancesChange(Set<String> enabledInstances, List<String> changed
-      segmentToOfflineInstancesMap.put(segment, offlineInstances);
+      Map<String, String> idealStateInstanceStateMap = entry.getValue();
+      Map<String, String> externalViewInstanceStateMap =
+          externalViewAssignment.getOrDefault(segment, Collections.emptyMap());
+      // Sort the online instances for replica-group routing to work. For multiple segments with the same online
+      // instances, if the list is sorted, the same index in the list will always point to the same instance.
+      Set<String> onlineInstances = new TreeSet<>();
       for (Map.Entry<String, String> instanceStateEntry : externalViewInstanceStateMap.entrySet()) {
         String instance = instanceStateEntry.getKey();
-
         // Only track instances within the ideal state
         // NOTE: When an instance is not in the ideal state, the instance will drop the segment soon, and it is not safe
         // to query this instance for the segment. This could happen when a segment is moved from one instance to
         // another instance.
         if (!idealStateInstanceStateMap.containsKey(instance)) {
           continue;
         }
-
         String externalViewState = instanceStateEntry.getValue();
         // Do not track instances in ERROR state
         if (!externalViewState.equals(SegmentStateModel.ERROR)) {
-          instanceToSegmentsMap.computeIfAbsent(instance, k -> new ArrayList<>()).add(segment);
-          if (externalViewState.equals(SegmentStateModel.OFFLINE)) {
-            offlineInstances.add(instance);
-          } else {
+          if (SegmentStateModel.isOnline(externalViewState)) {
             onlineInstances.add(instance);
           }
+        } else {
+          // Segment with error state instance should be considered old.
+          newSegmentStateMap.remove(segment);
         }
       }
-
-      // Sort the online instances for replica-group routing to work. For multiple segments with the same online
-      // instances, if the list is sorted, the same index in the list will always point to the same instance.
-      if (!(externalViewInstanceStateMap instanceof SortedMap)) {
-        onlineInstances.sort(null);
-        offlineInstances.sort(null);
-      }
-    }
-  }
-
-  /**
-   * Calculates the enabled ONLINE/CONSUMING instances for the given segment, and updates the unavailable segments (no
-   * enabled instance or all enabled instances are in ERROR state).
-   */
-  @Nullable
-  private List<String> calculateEnabledInstancesForSegment(String segment, List<String> onlineInstancesForSegment,
-      Set<String> unavailableSegments) {
-    List<String> enabledInstancesForSegment = new ArrayList<>(onlineInstancesForSegment.size());
-    for (String onlineInstance : onlineInstancesForSegment) {
-      if (_enabledInstances.contains(onlineInstance)) {
-        enabledInstancesForSegment.add(onlineInstance);
+      SegmentState state = newSegmentStateMap.get(segment);
```

Comment:
Let's rename this to `newSegmentState`

Reply:
this could be old segment too when it gets retired by clock